src/Compiler/Checking/CheckDeclarations.fs

```diff
@@ -3585,14 +3640,219 @@ module EstablishTypeDefinitionCores =
+
+                                | None -> tcFieldsAndSpreads fields i tpenv fieldsAndSpreads
+
+                            | SynFieldOrSpread.Spread (SynTypeSpread (ty = ty; range = m)) :: fieldsAndSpreads ->
+                                checkLanguageFeatureAndRecover g.langVersion LanguageFeature.RecordSpreads m
+
+                                let (spreadSrcTy, tpenv), error =
+                                    try TcType cenv NoNewTypars CheckCxs ItemOccurrence.UseInType WarnOnIWSAM.Yes envinner tpenv ty, false with
+                                    | RecoverableException e ->
+                                        errorRecovery e ty.Range
+                                        (g.obj_ty_ambivalent, tpenv), true
+
+                                if error || isRecdTy g spreadSrcTy || isAnonRecdTy g spreadSrcTy then
+                                    // It seems very likely that there's a better/existing way of doing this.
+                                    let spreadSrcTy =
+                                        tryAppTy g spreadSrcTy
+                                        |> ValueOption.map (fun (tcref, tinst) ->
+                                            let tinst =
+                                                tinst
+                                                |> List.map (fun ty ->
+                                                    tryDestTyparTy g ty
+                                                    |> ValueOption.map (fun typar ->
+                                                        let typars, _, _ = FreshenAndFixupTypars g m TyparRigidity.Flexible [] [] [typar]
+                                                        mkTyparTy (List.head typars))
+                                                    |> ValueOption.orElseWith (fun () ->
+                                                        let tryDestMeasureTy g ty =
+                                                            match stripTyEqns g ty with
+                                                            | TType_measure m -> ValueSome m
+                                                            | _ -> ValueNone
+
+                                                        tryDestMeasureTy g ty
+                                                        |> ValueOption.bind (function Measure.Var typar -> ValueSome typar | _ -> ValueNone)
+                                                        |> ValueOption.map (fun typar ->
+                                                            let typars, _, _ = FreshenAndFixupTypars g m TyparRigidity.Flexible [] [] [typar]
+                                                            TType_measure (Measure.Var (List.head typars))))
+                                                    |> ValueOption.defaultValue ty)
+
+                                            TType_app (tcref, tinst, g.knownWithoutNull))
+                                        |> ValueOption.defaultValue spreadSrcTy
```

Comment:
There _must_ be a better way of freshening up the type parameter here.

Reply:
Cannot this follow what "FreshenMethInfo" does, but on the level of appTy instead?
Doing a match on tryAppTy result, and then FreshenAppTy (a new function added to constraint solver). Working on the original declared typars of the tyconRef (like FreshenTypeInst g range0 (tcref.Typars range0) 

I think the measure part is not needed, at least I have not seen it being used elsewhere (the type argument  will then be a ttype_var pointing to a measure imo, instead of a direct ttype_measure - so it will get freshened up as well).


After you get freshtypars from FreshenTypeInst .. tcref.Typars ;; SolveTyparsEqualTypes from constraint solver will set the linkage between your new freshened typars and the original tinst.